Add upload-symbols command for Crashes

bryukov opened the pull request:
Implemented upload-symbols command in accordance to the specification - "-s", "-x" and "-m" switches are supported.
Command execution example: 
![upload-symbols](https://cloud.githubusercontent.com/assets/25530829/22943674/7870e5ce-f2ff-11e6-8936-627d36ae7d73.PNG)

Help:
![help](https://cloud.githubusercontent.com/assets/25530829/22943685/833c0cd6-f2ff-11e6-9727-d6dbe846bec3.png)

Comments and suggestions are welcome!

olegd commented:
Looks good. Please add tests.

christav commented on src/commands/crashes/upload-symbols.ts:
```diff
@@ -0,0 +1,420 @@
+  private detectSymbolsFsEntryType(filePath: string, fsEntryStats: Fs.Stats): SymbolFsEntryType {
+    if(fsEntryStats.isDirectory()){
+      // Check if it is a dSYM or xcarchive directory
+      switch(this.getLowerCasedFileExtenstion(filePath)){
```
`Extension` is spelled wrong.

christav commented on src/commands/crashes/upload-symbols.ts:
```diff
@@ -0,0 +1,420 @@
+    return SymbolFsEntryType.Unknown;
+  }
+
+  private getLowerCasedFileExtenstion(filePath: string): string{
```
`Extension` not `Extenstion`. Space before opening `{`

christav commented on src/commands/crashes/upload-symbols.ts:
```diff
@@ -0,0 +1,420 @@
+      desiredStatus,
+      cb, 
+    )).catch((error: any) => {
+        return Promise.reject(`failed to finalize the symbol upload with status '${desiredStatus}': ${this.getErrnoExceptionString(error)}`);
```
Throw an exception here instead of calling Promise.reject. That way handling the failure will consistently be an Error object.

christav commented on src/commands/crashes/upload-symbols.ts:
```diff
@@ -0,0 +1,420 @@
+}
+
+class UploadSymbolsError extends Error {
+  constructor(public errorCode: ErrorCodes, message: string){
```
Shouldn't you be storing this errorCode somewhere?

bryukov replied:
Yes, it is stored, I have used some syntactic sugar here - e.g. https://jsfiddle.net/gawg9ru5/
It is called "Parameter properties" in [TS handbook](https://www.typescriptlang.org/docs/handbook/classes.html)

bryukov commented:
Fixed pull request comments, moved UploadSymbolsError class to a separate file, fixed some of the source code comments;

@christav Can we add a TSLint configuration to the repository? It may really help us to detect formatting issues in future.

bryukov commented on src/commands/crashes/category.txt:
```diff
@@ -0,0 +1 @@
+Commands related to Crashes
```
@anlinde, @olegd Could you, please, provide the help message for the "crashes" command category?

anlinde replied:
For now the text should be: `Command to upload symbols to get symbolicated crash reports on Mobile Center`

christav replied:
@anlinde this file contains the help for the category, not just the one command. Is there better text for the set of commands as a whole? Or are you only ever going to have the one command?

anlinde replied:
We'll modify this text once we'll add more commands. But in the next 6 months it is unlikely we'll add more

anlinde commented:
The help text doesn't look good to me, especially for the symbol option. It describes what the CLI needs to do depending on the data provided by the user, it doesn't describe what the user can do

anlinde commented:
Help text suggestions:

`--symbol`: Path to a dSYM package, a directory containing dSYM packages, or a zip file containing the dSYM packages.

`--xcarchive`: Path to a xcarchive package

`--sourcemap`: Path to a React Native sourcemap file. Only supported in combination with `--symbol` or `--xcarchive`

bryukov commented:
Added test for the implemented command
